packages/adapters-library/src/adapters/beefy/products/moo-token/beefyMooTokenAdapter.ts

```diff
@@ -0,0 +1,316 @@
+  MovementsByBlock,
+  PositionType,
+  ProtocolAdapterParams,
+  ProtocolDetails,
+  ProtocolPosition,
+  ProtocolTokenTvl,
+  Underlying,
+  UnderlyingReward,
+  UnwrapExchangeRate,
+  UnwrapInput,
+} from '../../../../types/adapter'
+import { Erc20Metadata } from '../../../../types/erc20Metadata'
+import { Protocol } from '../../../protocols'
+
+type Metadata = Record<
+  string,
+  {
+    protocolToken: Erc20Metadata
+    underlyingTokens: Erc20Metadata[]
+  }
+>
+
+const chainIdMap: Record<Chain, string> = {
+  [Chain.Ethereum]: 'ethereum',
+  [Chain.Optimism]: 'optimism',
+  [Chain.Bsc]: 'bsc',
+  [Chain.Polygon]: 'polygon',
+  [Chain.Fantom]: 'fantom',
+  [Chain.Base]: 'base',
+  [Chain.Arbitrum]: 'arbitrum',
+  [Chain.Avalanche]: 'avax',
+  [Chain.Linea]: 'linea',
+}
+
+export class BeefyMooTokenAdapter
+  implements IProtocolAdapter, IMetadataBuilder
+{
+  productId = 'moo-token'
+  protocolId: Protocol
+  chainId: Chain
+  helpers: Helpers
+
+  private provider: CustomJsonRpcProvider
+
+  adaptersController: AdaptersController
+
+  constructor({
+    provider,
+    chainId,
+    protocolId,
+    adaptersController,
+    helpers,
+  }: ProtocolAdapterParams) {
+    this.provider = provider
+    this.chainId = chainId
+    this.protocolId = protocolId
+    this.adaptersController = adaptersController
+    this.helpers = helpers
+  }
+
+  /**
+   * Add your protocol details
+   */
+  getProtocolDetails(): ProtocolDetails {
+    return {
+      protocolId: this.protocolId,
+      name: 'Beefy',
+      description: 'Beefy defi adapter',
+      siteUrl: 'https://beefy.com',
+      iconUrl: 'https://beefy.com/icons/icon-96x96.png',
+      positionType: PositionType.Supply,
+      chainId: this.chainId,
+      productId: this.productId,
+      assetDetails: {
+        type: AssetType.StandardErc20,
+      },
+    }
+  }
+
+  @CacheToFile({ fileKey: 'protocol-token' })
+  async buildMetadata(): Promise<Metadata> {
+    type ApiVault = {
+      id: string;
+      status: "active" | "eol";
+      earnedTokenAddress: string;
+      chain: string;
+      platformId: string;
+      token: string;
+      tokenAddress: string;
+      earnedToken: string;
+      isGovVault?: boolean;
+      bridged?: object;
+      assets?: string[];
+    };
+    type ApiToken = {
+      type: string,
+      id: string,
+      symbol: string,
+      name: string,
+      chainId: string,
+      oracle: string,
+      oracleId: string,
+      address: string,
+      decimals: number
+    }
+    type ApiTokenList = {
+      [chainId: string]: {
+        [tokenId: string]: ApiToken;
+      };
+    }
+    type ApiBoost = {
+      id: string,
+      poolId: string,
+    }
+    const vaults = await (await fetch('https://api.beefy.finance/vaults')).json() as ApiVault[]
+    const tokens = await (await fetch("https://api.beefy.finance/tokens")).json() as ApiTokenList
+    const boosts = await (await fetch("https://api.beefy.finance/boosts")).json() as ApiBoost[]
```

Comment:
Hi sir, thank you for your review 🙏

So, that's a difficult one.

Beefy do have contract factories so we can pull some of that info from on-chain calls, but:
- Beefy factories are permissionless so using on chain calls only we risk including a vault that wasn't created by us.
- There is no official, on chain registry
- there is some data in those apis that is not available on chain like the strategy type and underlying protocol (needed for unwrap)

If you prefer, I can pull the same data from github config files directly, which would reduce the risk of it going down, but those files are more likely to change format.

Since the metadata is cached on a file, and the api is less likely to change than the GitHub files, I think this is the most stable and safest option, but I'm open to suggestions

Reply:
👋 Hey hey,

Thanks for the quick updates, and sorry for my late reply – was out since last week.

Great job on the adapters, seriously impressive stuff! 👏

Totally get the point on using APIs for metadata instead of on-chain calls. We've seen them used a few times for MRs. Discussed with a colleague today, and we're thinking of going ahead with merging adapters that use centralised APIs for metadata but all other methods must be do onchain so we can verify numbers. We'll tackle any API issues as they come up. 

If you've got any other ideas or suggestions, e.g. our metadata strategy I'm all ears.

Thanks again :)

Reply:
Hey thanks for your answer sir!

Yeah I think that's a good tradeoff between simplicity and stability.

Tbh so far, this is the greatest adapter system I've used. Miles ahead of anything we wrote for LRT points accounting. Congrats!

I'll go ahead and resolve this comment and fix all other comments asap!
Thx again